Add CMake support

lilianmoraru opened the pull request:
There are a few things to note here:
- This does not invalidate the `qmake` project files and builds, these can still work in parallel.
- If `ccache` or `sccache` is detected in the system, it will automatically be used
- If a faster linker is supported, it will be used(`LLD Linker` on `x86_64` with Clang, `Gold Linker` otherwise. If none supported, the default linker will be used)
- `stacer-core` is linked statically since it is a small codebase.
- Install Prefix was fixed(compared to the `qmake` version which always installs `stacer-core` in `/usr/lib`). You can use `-DCMAKE_INSTALL_PREFIX=<path>` or `make DESTDIR=<path> install`. Default prefix is `/usr/local` - that means, without passing anything, `make install` will put the binary here: `/usr/local/bin/stacer`
- `make install` will not install anything for `Debug` builds since no one should(probably by mistake) install unoptimized builds.
- Works in any IDE

lilianmoraru committed:
Add CMake support

lilianmoraru commented:
`cxxbasics` is copied from here: https://github.com/lilianmoraru/cxxbasics .
That project is used in production at the company I work.

lilianmoraru commented:
Btw, best way to try this is with `clang 4.0+` + `ccache` + `ninja`:
```
$ mkdir build && cd build
$ cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=clang++ -G Ninja ..
$ ninja
$ output/bin/stacer
```

oguzhaninan commented:
Thank you very much for your advice and support. But when I run the `ninja` command, I get this error.

![deepinscreenshot20170831213921](https://user-images.githubusercontent.com/16149251/29939560-e72f9f7e-8e94-11e7-936c-22576a0c80b1.png)

lilianmoraru commented:
Probably the `AUTOUIC` step failed, or something like this.
Could you please remove the contents of the `build` folder and then run again the `cmake` command + `ninja -v` and then paste the output of both of them, here?

oguzhaninan commented:
[cmake_out.txt](https://github.com/oguzhaninan/Stacer/files/1268457/cmake_out.txt)
[ninja_out.txt](https://github.com/oguzhaninan/Stacer/files/1268458/ninja_out.txt)

lilianmoraru commented:
@oguzhaninan I use `CMake 3.9.1`, but I can reproduce this issue with the default `cmake` available in `Ubuntu 16.04`: `CMake 3.5.1`...
It seems to be a known problem with this version of `CMake` when doing out-source builds...
Would you prefer me to add a few more `include_directories` to the build directory of `CMake` or just recommend to use a newer version of `CMake`?

Meanwhile you can play with it in-source:
```
$ cmake -G Ninja .
$ ninja
$ output/bin/stacer
```

oguzhaninan commented:
It worked. Thanks. Are all files in the [repository](https://github.com/lilianmoraru/cxxbasics) required?

lilianmoraru commented:
@oguzhaninan Except for `ClCache` which was never implemented and `Asserts`(which is used only by cxxbasics extensions), everything is pretty much required if you want faster builds and more reliable linkers(ex: making sure `LLD` is run only on `x86_64` - the only target where it is considered production ready).
`cxxbasics` takes ~80 KB right now.

As for the out-of-tree issue - I understand you want to use the `cmake` that comes with the system.
In that case I would like to add more `include_directories` and fix the issue for older versions of `cmake`.

oguzhaninan commented:
There is no problem for me. But it may be better for other developers. If you can do it, it will be fine but if not, I will merge.

lilianmoraru commented:
I removed some of the unnecessary files and fixed that out-of-tree issue: https://github.com/oguzhaninan/Stacer/pull/128/files#diff-0ac15bb75ead9dd03bd1b66ce588e75eR19
Now the initial commands(`mkdir build`...) I mentioned should work.

probonopd commented:
Seems like the builds fail now: https://travis-ci.org/probonopd/Stacer/builds/271097981

lilianmoraru commented:
@probonopd This does not change anything in the source code or file structure.
The issue I see in the link you provided is exactly the same issue you reported 7 days ago: https://github.com/oguzhaninan/Stacer/issues/123

lilianmoraru commented:
If you want to use `cmake` instead of `qmake` - instead of this:
```
qmake PREFIX=/usr
make -j$(nproc)
make INSTALL_ROOT=appdir install
```
run this:
```
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=appdir/usr .
cmake --build . --target install -- -j $(nproc)
```

probonopd commented:
Thanks @lilianmoraru, sorry, my mistake.

probonopd commented:
Getting a new error now:

```
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):

  CMake 3.4 or higher is required.  You are running version 3.2.2
```

Is it possible to downgrade the minimal required CMake version to 3.2.2? Then I could make continuous builds on Travis CI.

lilianmoraru commented:
@probonopd `CMAKE_CXX_COMPILER_LAUNCHER` used to accelerate clean builds with `ccache`/`sccache` requires `CMake 3.4`(you have `3.5` in `Ubuntu 16.04` which I personally consider pretty old).
I initially had it at `3.8` because I hate when people use old software(example: newer `CMake` adds features and fixes but also makes builds more reliable by making sure that when you update a system header or library on which the build depends, that it runs a rebuild, as an example).

To be sincere, I prefer people to update their software, but I could make it disable features if the version of `CMake` is too old.

probonopd commented:
Yes, please do that, otherwise I cannot compile it on Ubuntu 14.04 which is a still supported LTS version and is needed for AppImage compatibility with older systems.

lilianmoraru commented:
@probonopd Lowered to `3.1` in: https://github.com/oguzhaninan/Stacer/pull/141

vinyasns commented:
$ cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=/home/vinyas/Software/Qt-5.6/5.6/gcc_64/bin -G Ninja ..
-- [cxxbasics][✓] CMAKE_C_COMPILER(/usr/bin/cc)'s linker is set to: GNU gold linker
-- [cxxbasics][✓] CMAKE_CXX_COMPILER(/usr/bin/c++)'s linker is set to: GNU gold linker
-- [cxxbasics][✓] Compiler cache tool "ccache" set for the C compiler
-- [cxxbasics][✓] Compiler cache tool "ccache" set for the CXX compiler
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/Qt5/Qt5Config.cmake:26 (find_package):
  Could not find a package configuration file provided by "Qt5Charts" with
  any of the following names:

    Qt5ChartsConfig.cmake
    qt5charts-config.cmake

  Add the installation prefix of "Qt5Charts" to CMAKE_PREFIX_PATH or set
  "Qt5Charts_DIR" to a directory containing one of the above files.  If
  "Qt5Charts" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:16 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/vinyas/Desktop/Stacer/build/CMakeFiles/CMakeOutput.log".

I get this error while running cmake

lilianmoraru commented:
@vinyasns That's because the example in the `README.md` is not very good.
I would rather recommend:
```
$ cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=/home/vinyas/Software/Qt-5.6/5.6/gcc_64 ..
$ cmake --build . -- -j $(nproc)
```

Note that you need to use `gcc_64`, not `gcc_64/bin`, or just export the path to `qmake`( `export PATH="/home/vinyas/Software/Qt-5.6/5.6/gcc_64/bin:$PATH"` ) - in which case you don't need `CMAKE_PREFIX_PATH`.

vinyasns commented:
The same error occurs again

lilianmoraru commented:
@vinyasns Do you have Qt 5.6 commercial edition? Because Charts were not released in the open-source version until 5.7, with the move to GPLv3/LGPLv3.
If it is the open-source version, consider [installing a newer version](https://download.qt.io/official_releases/qt/5.9/5.9.1/) of Qt(preferably Qt 5.9+).

oguzhaninan commented:
@vinyasns What is your distribution.

lilianmoraru commented:
@oguzhaninan I don't think it is a distribution issue.
It's simply because the open-source version of Qt 5.6(the last GPLv2/LGPLv2.1 version) was not coming with pre-built QtCharts(it was commercial-only).

oguzhaninan commented:
@lilianmoraru There is no relationship to the distribution, but these packages need to be installed. 
`qt59base qt59imageformats qt59svg qt59charts-no-lgpl qt59tools` That's why I asked.  I experienced the problem and I solved it.

oguzhaninan commented:
@vinyasns You can follow the steps in the [Travis file](https://github.com/oguzhaninan/Stacer/blob/native/.travis.yml).

vinyasns commented:
I'm using open source version of Qt.
And Ubuntu 16.04.
@oguzhaninan I'll install those packages and try

oguzhaninan commented:
For Ubuntu 16.04
Run
- `sudo add-apt-repository ppa:beineri/opt-qt59-xenial && sudo apt-get update`
- `sudo apt-get install -y qt59base qt59imageformats qt59svg qt59charts-no-lgpl qt59tools`
- add to *~/.bashrc* file `source /opt/qt59/bin/qt59-env.sh`

vinyasns commented:
@oguzhaninan <a href="https://github.com/oguzhaninan/Stacer/blob/native/.travis.yml">Travis File</a> helped to build.
